### PR TITLE
Show Notification if downloading pdf is forbidden

### PR DESCRIPTION
**Pull request by JanAckermann**

### CODE REVIEW

**Comment by JanAckermann** [comment]
@mrow4a Can you please have a look at this, I am not sure if it is the right way to throw an exception, anyways I can't figure out how to get this in touch with l10n

**Review by mrow4a — COMMENTED** [review]
I will test drive it today

**Comment by JanAckermann** [comment]
> I will test drive it today

Thanks but the tests failing because of this lines

```
		'@phan-var \OCA\Files_Sharing\SharedStorage $storage';
		/** @var \OCA\Files_Sharing\SharedStorage $storage */
		$share = $storage->getShare(); 
```

Do you have any Idea to get rid of this problem?

File: lib/Controller/DisplayController.php [hunk 1]
```diff
@@ -55,4 +58,33 @@ public function showPdfViewer() {
...
+		$storage = $fileInfo->getStorage();
+		$canDownload = true;
+
+		if (!$storage->instanceOfStorage('OCA\Files_Sharing\SharedStorage')) {
+			return new JSONResponse(['canDownload' => $canDownload]);
+		}
+
+		/** @var \OCA\Files_Sharing\SharedStorage $storage */
+		'@phan-var \OCA\Files_Sharing\SharedStorage $storage';
+		$share = $storage->getShare();
```
Comment (mrow4a) [inline lib/Controller/DisplayController.php:79]: @phil-davis I had a look, but I am also puzzled why we get into problem here https://drone.owncloud.com/owncloud/files_pdfviewer/860/5/3  . Are we missing something in makefile etc?

File: appinfo/routes.php [hunk 2]
```diff
@@ -12,4 +12,5 @@
 
 return ['routes' => [
 	['name' => 'display#showPdfViewer', 'url' => '/', 'verb' => 'GET'],
+	['name' => 'display#canDownload', 'url' => '/candownload', 'verb' => 'GET'],
```
Comment (mrow4a) [inline appinfo/routes.php:15]: great idea! even better would be to, if possible, detect if this file is a share, and query for "attribute -> canDownload" via https://doc.owncloud.org/server/8.2/developer_manual/core/ocs-share-api.html#get-information-about-a-known-share .

Comment (JanAckermann) [reply appinfo/routes.php:15]: Agree, but that means, we would need a share id in the previewplugin js, what i think we don't have

**Comment by phil-davis** [comment]
I added commit "Tell phan to look in files_sharing/lib for classes" - now `make test-php-phan` passes.

But `make test-php-phpstan` fails with:
```
 ------ ---------------------------------------------------------------------------------------------------------------- 
  Line   lib/Controller/DisplayController.php                                                                            
 ------ ---------------------------------------------------------------------------------------------------------------- 
  78     Expression "'@phan-var \\OCA\\Files_Sharing\\SharedStorage $storage'" on a separate line does not do anything.  
 ------ ---------------------------------------------------------------------------------------------------------------- 
                                                                                                                        
 [ERROR] Found 1 error                               
```

Next challenge is how to tell `phpstan` to ignore the `phan-var` code!

**Comment by JanAckermann** [comment]
> I added commit "Tell phan to look in files_sharing/lib for classes" - now `make test-php-phan` passes.
> 
> But `make test-php-phpstan` fails with:
> 
> ```
>  ------ ---------------------------------------------------------------------------------------------------------------- 
>   Line   lib/Controller/DisplayController.php                                                                            
>  ------ ---------------------------------------------------------------------------------------------------------------- 
>   78     Expression "'@phan-var \\OCA\\Files_Sharing\\SharedStorage $storage'" on a separate line does not do anything.  
>  ------ ---------------------------------------------------------------------------------------------------------------- 
>                                                                                                                         
>  [ERROR] Found 1 error                               
> ```
> 
> Next challenge is how to tell `phpstan` to ignore the `phan-var` code!

thanks, solved

File: tests/unit/Controller/DisplayControllerTest.php [hunk 3]
```diff
@@ -12,8 +12,10 @@
 
 use OCP\AppFramework\Http\ContentSecurityPolicy;
 use OCP\AppFramework\Http\TemplateResponse;
+use OCP\AppFramework\Http\JSONResponse;
 use OCP\IRequest;
 use OCP\IURLGenerator;
+use phpDocumentor\Reflection\Types\This;
```
Comment (phil-davis) [inline tests/unit/Controller/DisplayControllerTest.php:18]: my IDE tells me this is unused

**Review by mrow4a — APPROVED** [review]

**Comment by mmattel** [comment]
Under which circumstances a download is forbidden?
From where do we know that?
If a user can see a file how can it be distingusihed that he cant download it?

Could you please make some comments in the PR description? thanks

**Comment by JanAckermann** [comment]
@mmattel description added

**Comment by mmattel** [comment]
@JanAckermann great, thanks.

Q: `This shared file does not have download permission...` and `...shared PDF is proteced by secure view`
Why didnt you mention in the result text that this could probably be proteced by secure view?

Someting like:
`This shared file does not have download permission and is possibly proteced by secure view, please contact the owner of the file for granting permission or use a different viewer`.
As a user I would then know the cause and it is easier to ask for a grant knowing what the possible cause is...

**Comment by JanAckermann** [comment]
@mmattel Good point!

**Comment by mmattel** [comment]
Maybe you want to add a small fix PR for that text :smile:
If you add me as reviewer, I will approve asap (when I come back from IKEA, if I survive that...).

**Comment by JanAckermann** [comment]
@mmattel  tommorow :) good luck at IKEA :D

**Comment by hurradieweltgehtunter** [comment]
> This apply when a shared PDF is proteced by secure view.

Just had the feedback by a oc.online customer that this happened here too but without SecureView (not available for oc.online). Does the preflight check also work for non-SecureView setups?

**Comment by JanAckermann** [comment]
@hurradieweltgehtunter depends, in which cases does this happen?

**Comment by hurradieweltgehtunter** [comment]
Quoting customer:
"This is the error my colleague got when he clicked on a PDF that he did not have download rights to. Could you please help us figure out how to resolve this issue and prevent this error from occurring? Is there a way to make PDF files viewable without letting the user download them?!
grafik

I was able to reproduce it by disabling all privileges in the sharing dialogue:
grafik

**Comment by JanAckermann** [comment]
> Quoting customer:
> "This is the error my colleague got when he clicked on a PDF that he did not have download rights to. Could you please help us figure out how to resolve this issue and prevent this error from occurring? Is there a way to make PDF files viewable without letting the user download them?!
> grafik
> 
> I was able to reproduce it by disabling all privileges in the sharing dialogue:
> grafik

It should, but I am not pretty sure, can you checkout the issue and test it, as I am in vacations now? Would be great!
Merry Xmas

**Comment by jnweiger** [comment]
Tested with richdocuments and secureview enabled:

image
image

This seems to be the intended behaviour now.

**Pull request merged**